index.mjs

```diff
@@ -171,38 +195,81 @@ export default class MediaMixin extends Componentry.Module {
       }
     })
 
-    router.put('/media/upload/*',async (req,res)=>{
+    router.put('/media/upload/*',async (req,res) => {
       if (!req.account) return res.status(401).send();
 
       try {
-        let mediaItem = await this.collection.findOne({_id:req.params[0]},);
+        let mediaItem = await this.collection.findOne({_id:req.params[0]});
         if (!mediaItem) return res.status(400).send(`${req.params[0]} has not been staged`);
         let buffer = req.files.file.data;
         let fileType = req.files.file.mimetype;
         let file = `${mediaItem._id}.${fileType.split('/')[1]}`;
 
         // Normalize images into PNG and capture initial crop spec
-        if (fileType.startsWith("image/")) {
+        if (fileType.startsWith('image/')) {
           fileType = 'image/png';
-          file = `${mediaItem._id}.png`;
+
+          if (Object.keys(req.query).length === 0) {
+            const thumbnails = [{scale: '60,60,cover'}]
+
+            for (const thumbnail of thumbnails) {
+              let specForThumbnail = await this.storage.getSpec(mediaItem._id, thumbnail);
+              let bufferThumbnail = await sharp(buffer,{failOnError: false});
+              bufferThumbnail = await specForThumbnail.process(bufferThumbnail);
+              await this.storage.putImage(mediaItem._id, specForThumbnail.path, fileType, bufferThumbnail, false);
+              await this.storage.commitThumbnailOnDb(mediaItem._id, thumbnail)
+            }
+          }
 
           let spec = await this.storage.getSpec(mediaItem._id, req.query);
           buffer = await sharp(buffer,{failOnError: false});
           buffer = await spec.process(buffer);
+          file = spec.path
         }
-        let result = await this.storage.putImage(mediaItem._id,file,fileType,buffer);
+
+        await this.storage.putImage(mediaItem._id, file, fileType, buffer);
         res.json({});
       } catch (e) {
         console.error('/media/upload/* error:', e);
         res.status(500).send();
       }
     });
+
+    router.put('/media/stage/*',async (req,res) => {
+      try {
+        if (!req.account) return res.status(401).send();
+        const server = req.params[0] || 'aws'
+
+        console.log('Request body:', req.body)
+
+        switch (server) {
```

Comment:
I was trying to avoid having the image known both to the database and the storage system. This requires keeping them in sync. The name of the image file has the account from the bucket string. It has an id and rendering variants for each named target. It is modified with '.image' so one root image can be rendered differently for the same parent object. See Manifest.mjs. It defines default specs for each named target, like FB (facebook), GG (google), icon (for an icon, which should not have a hardcoded spec. The user may modify the default by dragging the crop window and/or zooming. The custom spec than overrides the default. Each variant is saved as a different image with the same ID root, for efficiency.